test/functional/p2p_eviction.py

```diff
@@ -0,0 +1,141 @@
+            current_peer += 1
+
+
+        #The next 4 peers send us a tx, protecting them from eviction
+        for i in range(4):
+            txpeer = node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+            txpeer.sync_with_ping(timeout=5)
+
+            prevtx = node.getblock(node.getblockhash(i+1), 2)['tx'][0]
+            rawtx = node.createrawtransaction(
+                inputs=[{'txid': prevtx['txid'], 'vout': 0}],
+                outputs=[{node.get_deterministic_priv_key().address: 50 - 0.00125}],
+            )
+            sigtx = node.signrawtransactionwithkey(
+                hexstring=rawtx,
+                privkeys=[node.get_deterministic_priv_key().key],
+                prevtxs=[{
+                    'txid': prevtx['txid'],
+                    'vout': 0,
+                    'scriptPubKey': prevtx['vout'][0]['scriptPubKey']['hex'],
+                }],
+            )['hex']
+            txpeer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+            self.protected_peers.add(current_peer)
+
+        # The next 8 peers have faster pings, which will usually protect them from eviction.
+        for _ in range(8):
+            fastpeer = node.add_p2p_connection(P2PInterface())
+            current_peer += 1
+            wait_until(lambda: "ping" in fastpeer.last_message, timeout=10)
+
+        # Make sure by asking the node what the actual min pings are
+        peerinfo = node.getpeerinfo()
+        pings = [[0 for i in range(2)] for j in range(len(node.p2ps))]
+
+        # After adding 21 peers, the next one hits the maxconnection limit and triggers the eviction mechanism.
+        node.add_p2p_connection(SlowP2PInterface())
+
+        # Usually the 8 fast peers are protected. In rare case of unreliable pings,
+        # one of the slower peers might have a faster min ping though.
+        for i in range(len(peerinfo)):
+            pings[i][0] = i
+            pings[i][1] = peerinfo[i]['minping'] if 'minping' in peerinfo[i] else 1000000
+
+        sorted_pings = sorted(pings, key = lambda x:x[1])
+
+        for i in range(8):
+            self.protected_peers.add(sorted_pings[i][0])
+
+        # One of the non-protected peers must be evicted. We can't be sure which one because
+        # 4 peers are protected via netgroup, which is identical for all peers,
+        # and the eviction mechanism doesn't preserve the order of identical elements.
+        self.evicted_peers = []
+        for i in range(len(node.p2ps)):
+            if(node.p2ps[i].is_connected == False ) :
+                self.evicted_peers.append(i)
+
+        self.log.info("Test that one peer was evicted")
+        assert_equal(len(self.evicted_peers), 1)
+        self.log.debug("{} evicted peer: {}".format(len(self.evicted_peers), set(self.evicted_peers)))
+
+        self.log.info("Test that no peer expected to be protected was evicted")
+        assert_equal(self.evicted_peers[0] in self.protected_peers, False)
```

Comment:
style nit: assertion here is about exclusion, not equality; I think the idiomatic assertion in the tests codebase would be
```suggestion
        assert self.evicted_peers[0] not in self.protected_peers
```

Reply:
done.